klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+    )
+    print(run_dir)
+
+    all_tables.insert(0, "main.drc")
+    all_tables.append("tail.drc")
+
+    gen_rule_deck_path = os.path.join(run_dir, "{}.drc".format(deck_name))
+    with open(gen_rule_deck_path, "wb") as wfd:
+        for f in all_tables:
+            with open(os.path.join(drc_dir, "rule_decks", f), "rb") as fd:
+                shutil.copyfileobj(fd, wfd)
+
+    return gen_rule_deck_path
+
+
+def get_top_cell_names(gds_path):
+    """
+    get_top_cell_names get the top cell names from the GDS file.
+
+    Parameters
+    ----------
+    gds_path : string
+        Path to the target GDS file.
+
+    Returns
+    -------
+    List of string
+        Names of the top cell in the layout.
+    """
+    layout = klayout.db.Layout()
+    layout.read(gds_path)
+    top_cells = [t.name for t in layout.top_cells()]
+
+    return top_cells
+
+
+def get_list_of_tables(drc_dir: str):
+    """
+    get_list_of_tables get the list of available tables in the drc
+
+    Parameters
+    ----------
+    drc_dir : str
+        Path to the DRC folder to get the list of tables from.
+    """
+    return [
+        os.path.basename(f).replace(".drc", "")
+        for f in glob.glob(os.path.join(drc_dir, "rule_decks", "*.drc"))
+        if "antenna" not in f
+        and "density" not in f
+        and "main" not in f
+        and "tail" not in f
+    ]
+
+
+def get_run_top_cell_name(arguments, layout_path):
+    """
+    get_run_top_cell_name Get the top cell name to use for running. If it's provided by the user, we use the user input.
+    If not, we get it from the GDS file.
+
+    Parameters
+    ----------
+    arguments : dict
+        Dictionary that holds the user inputs for the script generated by docopt.
+    layout_path : string
+        Path to the target layout.
+
+    Returns
+    -------
+    string
+        Name of the topcell to use in run.
+
+    """
+
+    if arguments["--topcell"]:
+        topcell = arguments["--topcell"]
+    else:
+        layout_topcells = get_top_cell_names(layout_path)
+        if len(layout_topcells) > 1:
+            logging.error(
+                "## Layout has mutliple topcells. Please determine which topcell you want to run on."
+            )
+            exit(1)
+        else:
+            topcell = layout_topcells[0]
+
+    return topcell
+
+
+def generate_klayout_switches(arguments, layout_path):
+    """
+    parse_switches Function that parse all the args from input to prepare switches for DRC run.
+
+    Parameters
+    ----------
+    arguments : dict
+        Dictionary that holds the arguments used by user in the run command. This is generated by docopt library.
+    layout_path : string
+        Path to the layout file that we will run DRC on.
+
+    Returns
+    -------
+    dict
+        Dictionary that represent all run switches passed to klayout.
+    """
+    switches = dict()
+
+    # No. of threads
+    thrCount = 2 if arguments["--thr"] is None else int(arguments["--thr"])
```

Comment:
should we default on the current number of CPU?

Reply:
@proppy We used to do that at the beginning, but after testing with klayout, the best was to use from 1 to 4 CPU maximum. If you have a 32 CPU machine it doesn't help klayout even in deep mode.

Reply:
@klayoutmatthias are there some documentation around CPU currency best practice for running klayout drc checks?